src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader.cs

```diff
@@ -266,24 +270,61 @@ private void ReadFirstToken(byte first)
             }
             else
             {
-                ConsumeValue(first);
+                _isSingleValue = true;
+                if (ConsumeValue(first))
+                {
+                    if (CurrentIndex >= (uint)_buffer.Length)
+                    {
+                        TokenType = JsonTokenType.Value;
+                        return true;
+                    }
+#if UseInstrumented
+                    throw new JsonReaderException($"Expected end of json after a single value but there is extra data within the payload. Last character read: '{(char)_buffer[CurrentIndex]}'.", _lineNumber, _position);
+#else
+                    JsonThrowHelper.ThrowJsonReaderException(ref this);
+#endif
+                }
+                return false;
             }
+            return true;
         }
 
         private bool ReadSingleSegment()
         {
             bool retVal = false;
 
             if (CurrentIndex >= (uint)_buffer.Length)
+            {
+                if (!_isSingleValue && _isFinalBlock)
```

Comment:
We invented isFinalBlock for parsers that cannot unambiguously tell between end of data and incomplete data. Is this the case here? Wouldn't it be possible to simply return either Incomplete or Done when the reader parses all data. This would avoid all the isFinalBlock checks that surely add overhead.